### DISCUSSION TITLE
Tags

### DISCUSSION THREAD

**Discussion by alvarix** [alvarix/FLogger#54] [Ideas]
# Title/Tag Model
I realize as thinking this through that I wasnt envisioning a tagging system.
Rather a title system that had some tagging properties.

'Tagging' is where multiple titles can be selected and displayed as one.
Whereas with traditional tags selecting mutiple tags brings up a collection of all the objects with those tags, in our system selecting multiple tags bring up a collection of entries across different notes.
Since all our notes use a standard chronological object format, merging several notes should be easy (I hope).
I'm going to read up on tagging a bit to see if I get other ideas, but wanted to get this down.

Dev Considerations:
If viewing a multi-merged note, is it possible to add a new entry from that interface, thus creating the same entry in every titled note being displayed?
- Is that useful to be able to have the same entry across multiple notes? I think so for individual note integrity.
- Will that scale?
1. Option A
- It will have to be reconciled on the front end to merge duplicates.
- Or, is it handled on the backend invisibly through a third meta note?

2. Option B - you cant add a note from a multi-note interface
- you can't add multiple tags, just a title (select from existing or new)
- way simpler to implement
- slightly less useful

This all starts with the assumption that each Title is has it's own corresponding text data file.
In my mind the word Note refers to the title with all of it's timestamped entries.
Both the terminology we are using in our components and your experimentation with text file data doesn't go in this direction.
So how data is saved my play a big part in architecting any tag or tag/title system.

I think we need to put our heads together about how the DB works as a prerequisite to tagging.

**Comment by alvarix** [comment]
Chad's response:

If I understand you correctly, this is how I'd describe what you want...

1) When you select one or more tags, you see a single, unified document view all entries of notes matching those tags. 
2) If you select multiple tags, the entries shown are from notes matching any of those tags, but doesn't have to match all of the selected tags.

Is that right?

3) I'm not sure what you mean by "a title system that had some tagging properties". Tags are basically words in note titles?

4) The same entry across multiple notes — I don't understand. Do you mean literally the same? Like I could edit one note and that would update the same entry in all notes that share that entry?

5) Adding a note while viewing a mutli-merged note ... I would think this creates a new note, new entry, and the new note has all the tags of the current view. I guess that means those tag words in the title?

6) I just don't think I get the full idea of what you're thinking. We should discuss more scenarios of how this would/could work. Not how it might be implemented, but literally just narrations of how someone uses it. 
What's an example of the list of tags/titles the user would select from? Or is it a search field and any words they type in are used to match words in note titles?
Once "tags" are selected, you see a view that looks like what? Maybe type in an example in a google doc.
From there you enter a new note. What would be most useful result?

**Comment by alvarix** [comment]
@chadcrume tx for your response. 
I've gotten so wrapped up in my thoughts about how this works, maybe you should describe how you'd envision it.

I've attached a crappy diagram below.

**Terms**
_**Entry**_: an individual timestamped object
_**Log**_: a collection of timestamped entries stored as an eponymous text file 
_**Title**_: Formerly know as _Tag_ (acts _tag-like_ in combo-log?)
_**Combo Log**_: Merged UI log of an array of titles

**User Story**
1. User opens FLogger, sees an omnibox for _search/create new entry_ and a list of tags.
2. Thru typing in teh box or manually clicking the tag list he can select one or more tags, which opens a UI log of all entries that have those tags.
3. If he wants to add an entry there are a few ways this can go:
a. The new entry asks for a singular tag.
b. The new entry accepts all tags in the current log and/or any other new or existing tags.

Version b is where the technical challenge I've been laboring to describe.
- Write entry to two logs
- merge those entries into one when viewing a combo log
For simplicity I think I'd prefer each entry to have a single 'tag' as a first step bcs work:utlity ratio.
I think some experience using a proptoype will clarify if having entries with multiple tags is worth the effort.


_Some of your questions:_
_When you select one or more tags, you see a single, unified document view all entries of notes matching those tags._
yes

_If you select multiple tags, the entries shown are from notes matching any of those tags, but doesn't have to match all of the selected tags.
Is that right?_
dont follow

_The same entry across multiple notes — I don't understand. Do you mean literally the same? Like I could edit one note and that would update the same entry in all notes that share that entry?_

To unravel my thinking - it goes back NValt - each note title is an eponymous text file. 
That software has a seperate tagging capability, which I dont use at all.
The omnibox makes create/search so efficient, that I dont want to muddy it with an additional taxonomy.
That's what has led me to try a blended title/tag approach. 
Maybe the term tag is causing harm and I should refer to them as simply _blended logs_.
Backing up my thinking just now has made me realize that for me it seems important not to add tags as additional categorization. I want the titles to be the only retrieval method.





![IMG_0512 (1)](https://github.com/alvarix/FLogger/assets/1242083/3f565a95-f766-4fb3-9d5f-b1c5349f432d)

**Reply by chadcrume** [reply]
Following up on this question...

C: If you select multiple tags, the entries shown are from notes matching any of those tags, but doesn't have to match all of the selected tags.
Is that right?
A: dont follow
C: You said that title = tag, and at first we should implement 1 tag per entry. In user story step 2 you mention that the user "can select one or more tags". So when the user selects more than one tag, say A and B, the combo log will show entries with title A and entries with title B.  Matching entries that have _both_ tag A and tag B doesn't make sense when each entry has only one tag. So our match shows entries that match _any_. That's all I was clarifying. Later when/if entries can have more than one tag, we can cover the match _both_/_all_ scenario.

**Reply by chadcrume** [reply]
@alvarix , this is all very clear. Thanks for detailing it so well. 

I get what you mean about title = tag. 
Question about title/tag allowed characters: 
I assume a title/tag can have any text character, including spaces. I just want to clarify, because I think there are two kinds of standard tag implementations out there in the world, those that consider tags as words (no spaces), and those that consider tags as unique strings (spaces allowed).

It makes sense to me to implement 1 title/tag per entry to start with. It might help focus on implementing your vision of logs and combo-logs gracefully merged in the UI.
Question about title/tag-based files: 
In your vision, you have one log per title/tag. And each log is a single file (or other datastore)? I still question this coupling a file to a single title/tag. It's complicated to explain. 
* The most common data modeling approach would be to have a single record in data for the entry, and have the title/tag(s) as properties of the entry. Different entries can have the same title/tags. This is fairly straightforward to match because title/tag values are unique identifiers themselves.
* Your model turns this inside out, making entries properties of title/tag(s). The title/tag file becomes a single object with entries as properties. And different files can have the same entries. Here each entry is an object and we match them probably using a GUID or URI property on that object. The entry text could change. Maybe the timestamp can be edited too, which would also need to sync across logs.
* What if you want to store all entries for 2022 separate from those for 2023? As a way to archive 2022. You could use a title/tag to collect all notes from 2022 into a single file that can be archived, but all those notes would still be duplicated across other title/tag log files. 
* What if you want to store super private entries separate from less private ones? You can't use the same title/tags or else super private notes would be available when you open the log for a title/tag.
* My gut tells me it will be worthwhile to think of a file as a collection of entries and title/tags. Each entry and each title/tag could be in any number of files. The UI shows and blends title/tags from any/all files the user opens at the same time. 
* But still, it might be simpler to start with 1 title/tag per log/file, 1 title/tag per entry, and see where we get. Focus on the merging scenarios required to get the UX right.  We can refactor the file model later.

So starting with **1 title/tag per log/file** and **1 title/tag per entry** , we don't have to worry yet about merging the same entry from multiple title/tag/log/files. We focus on merging all unique entries from multiple selected title/tag/log/files. Does that sound right?

**Comment by alvarix** [comment]
Chad you’re making a lot of interesting points.
I want to digest it - but first just clarify my idea.

A log is a collection of entries with the same title.
Entries can only have one title.
The reason for this - logs can be saved in a text file of that title,
log-name.txt

The only time you have more than one title is in the UI - a combined log.

There are reasons why a log may benefit for more than 1 title, but I mainly
see it as facilitating the
log-name.txt db system.

If that data system isn’t the way we go then I might be open to having
multiple titles (tags?) per note.







On Sun, Aug 20, 2023 at 08:28 C C C ***@***.***> wrote:

> @alvarix <https://github.com/alvarix> , this is all very clear. Thanks
> for detailing it so well.
>
> I get what you mean about title = tag.
> Question about title/tag allowed characters:
> I assume a title/tag can have any text character, including spaces. I just
> want to clarify, because I think there are two kinds of standard tag
> implementations out there in the world, those that consider tags as words
> (no spaces), and those that consider tags as unique strings (spaces
> allowed).
>
> It makes sense to me to implement 1 title/tag per entry to start with. It
> might help focus on implementing your vision of logs and combo-logs
> gracefully merged in the UI.
> Question about title/tag-based files:
> In your vision, you have one log per title/tag. And each log is a single
> file (or other datastore)? I still question this coupling a file to a
> single title/tag. It's complicated to explain.
>
>    - The most common data modeling approach would be to have a single
>    record in data for the entry, and have the title/tag(s) as properties of
>    the entry. Different entries can have the same title/tags. This is fairly
>    straightforward to match because title/tag values are unique identifiers
>    themselves.
>    - Your model turns this inside out, making entries properties of
>    title/tag(s). The title/tag file becomes a single object with entries as
>    properties. And different files can have the same entries. Here each entry
>    is an object and we match them probably using a GUID or URI property on
>    that object. The entry text could change. Maybe the timestamp can be edited
>    too, which would also need to sync across logs.
>    - What if you want to store all entries for 2022 separate from those
>    for 2023? As a way to archive 2022. You could use a title/tag to collect
>    all notes from 2022 into a single file that can be archived, but all those
>    notes would still be duplicated across other title/tag log files.
>    - What if you want to store super private entries separate from less
>    private ones? You can't use the same title/tags or else super private notes
>    would be available when you open the log for a title/tag.
>    - My gut tells me it will be worthwhile to think of a file as a
>    collection of entries and title/tags. Each entry and each title/tag could
>    be in any number of files. The UI shows and blends title/tags from any/all
>    files the user opens at the same time.
>    - But still, it might be simpler to start with 1 title/tag per
>    log/file, 1 title/tag per entry, and see where we get. Focus on the merging
>    scenarios required to get the UX right. We can refactor the file model
>    later.
>
> So starting with *1 title/tag per log/file* and *1 title/tag per entry* ,
> we don't have to worry yet about merging the same entry from multiple
> title/tag/log/files. We focus on merging all unique entries from multiple
> selected title/tag/log/files. Does that sound right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/alvarix/FLogger/discussions/54#discussioncomment-6772869>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJPHY5DUATGRAHO5YFJ2NDXWH7IRANCNFSM6AAAAAA3M7NDJA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 
Sent from Batmobile

**Comment by chadcrume** [comment]
Notes from conversation:


## Chad's follow up to-do

Will write up scenarios comparing contrasting title-tag log-files vs title-tag separate from log-file.


## Alvar's follow up to-do

Describe how the system will allow the user to open/close multiple title-tag-log-files.

> Omni box has list of all title-tag-log-files in a specific folder. Curious how to allow typing, auto-complete, and hitting enter to open a log, or allowing selecting one log and continue to type to select more than one.
> The idea of blending logs, need to write out use cases for this.

Alvar will think of how he sees these things being solved:

Be able to share some notes from a title-tag, and not other notes from the same title-tag.

Be able to store notes from 2022 separate from notes from 2023.

* What if you want to store all entries for 2022 separate from those for 2023? As a way to archive 2022. You could use a title/tag to collect all notes from 2022 into a single file that can be archived, but all those notes would still be duplicated across other title/tag log files.
* What if you want to store super private entries separate from less private ones? You can't use the same title/tags or else super private notes would be available when you open the log for a title/tag.

**Comment by alvarix** [comment]
Summarize some of my comments:

> For me the title, being the file with all the entries, they are in is a pretty fundamental part of the data portability.
> The way I envisioned the data file isn’t even Json if it can be avoided, but something that looks much more like a real note with natural language or symbols denoting all the data points. 
> 
> The reason being, so that app is just a convenience, but in no way necessary to interact with the data. You can just as easily open a note and add an entry just with a little bit more typing. That way if there’s ever a problem with the app it’s really frictionless, just keep the notes going with no Rosetta Stone necessary. 
> I totally get that this single feature might be so much of a burden that it’s not worth it.
> 
> I also get that there’s no such thing as private notes in the system, because you can’t encrypt that way.
> I’m curious what you think the advantage of a database based on text files is compared to a real database?

**Comment by alvarix** [comment]
Chad you and I clarified some conceptual differences in this app will solve our particular logging desires. 

- I prefer a very simple and straightforward tool in making and viewing single-title log files. 
- You prefer a tool that will will perform different views and other operations on tagged or otherwise organized blended entries log entries.

Please correct me if that's not accurate or oversimplified. 
Our next task is to see if we have to choose one direction over another, or if we can a flexible basis that will service both. 

The main obstacle is my vision of an almost mirror like data model, where the data itself are also the logs. 
Chad initially thinks that the data should be abstracted to allow flexibilty.

My challenge is to envision a data flexible scenario that also achieves the simplicity and redundancy (for lack of a better word, the lack of necessity of the app).

**Comment by alvarix** [comment]
To consider in use cases - are blended logs useful?

Chad's questions:

> 1. How to share some entries in a log and not others
> 2. Archive logs by date

1. I mentioned that I wasn't considering encryption due to possible incompatibility with text files. 
Though what chad describes here is another piece of metadata on an entry, that corresponds to the creation of custom views. At least by my model. Chad's model has this flexibility somewhat built in. 
In my model there might be a _private_ checkbox on each entry. 
On the log level there would be a _share_ button that makes the text file (or blend of text files) publicly accessible filtering out the private entries.

2. We discussed a chron that runs yearly renaming the logs and categorizing them into another folder, removing them from the default logs but providing UI to view archives.
Interesting questions here are what preferences can we provide (don't archive, archive in another way, etc. ).

**Comment by alvarix** [comment]
Two quick Blended Log use cases.
Are these compelling? Do they inspire any others?


**Mtk Rental Log & Mtk Service Log**
_View different service visits in context of what renter was there and what issues arose._

**Dev Project Log & Programming Language Log**
_View the different insights I've had in a programming language in context of what I was working on at the time, helping to understand why programming choices were made._

**Comment by chadcrume** [comment]
Alvar's POV is that a title, tag, and log file are all the same thing.

Chad's advanced scenarios:
1. One entry could have multiple tags
2. One tag could be separated into older archived entries, and current
3. Some entries from one tag could be shared without sharing all

Alvar's concern is keeping a super simple file format that doesn't require app logic to parse and be useful.

Maybe the difference of POVs is entirely based on the concept of tags being confused with log files and the root data container. We can move forward with Alvar's concept of entries in log files, and blending log files in the UI. Forget tags for now.

Core ethos: Entries are in files that are entirely portable and useful without an app.
Decisions: 
* Drop terminology of "tag" entirely for now. 
* Focus on logs that are single files with a single title containing multiple entries.

**Comment by chadcrume** [comment]
Resurrecting this discussion and updating it to our latest thinking.

# tags
 - sidebar: tags in current flog
 - on click it shows
    - all entries in flog with tag
    - all flogs with entries with tag

**Comment by chadcrume** [comment]
# thinking through implementation

## Stating the obvious: current flog content is already open in the app
The component for this will be in the sidebar of the current flog. So the entries and tags for the current flog are already downloaded from Dropbox and in app memory. 

## All other flogs need to be downloaded and scanned for tags
It is necessary to download all of the user's flog files from Dropbox in order to parse for tags matching the one the user clicked on.

 - There isn't an API for downloading multiple files in one call except filesDownloadZip, which returns a zip that our app would have to unzip into files and then open each file's contents.

## An alternative is to keep an index of tags
With pointers to flog entries containing them. This index would be updated every time a user saves any flog — the tags in that flog would be parsed and saved in the index with pointers back to the entries in that flog. We would save this index into the user's Dropbox folder, like settings

 - A **FlogTags** component will be in the side bar, a sibling of the ToC component, by tab or accordion or something. It will access the tags data via useDropboxFlogs.
 - A new composable **useTags** will provide an API for reading and updating the tag index in the app. It will encapsulate the local storage mechanism, state. But not the external storage. 
 - The existing composable **useDropboxFlogs** will useTags, and will integrate updating the tag index into the flog update events. It will re-export only the useTags state vars, not it's update API.
 - The existing composable **useDropboxFiles** will be updated to provide access to the tag index file from the user's Dropbox app folder. 

**useTags** will have a tag index rev prop, which will come from the Dropbox file rev value. When the local and the origin rev don't match, the local is updated from the origin, entirely overwritten, not merged.  So... 

### Flow 1: Load tag index
When Dropbox connection happens — basically the same code lines where the repo template files are managed — the tag index file content is downloaded. But only if needed. The current rev on the Dropbox files is compared to the last one saved in the app. This way we can avoid downloading the index file every time a user connects. 

This means that useDropboxFiles will need to have the current rev from useTags to do the comparison. Or, rather than passing in the rev, we can pass in a function which useDropboxFiles will use to perform the comparison check. useDropboxFiles calls it and passes in the new rev. The logic to compare it is delegated to the code using useDropboxFiles. 

### Flow 2: Update tag index
When the tag index is saved to Dropbox, the new rev has to be saved on the local index.  

This means that the useDropboxFiles save operation for the tag index file will need to return a rev to be saved in useTags. There's already a callback in the useDropboxFiles saveFileContent. So the new rev can be passed into that, and the code using useDropboxFiles can work with it from the callback.
